editor/src/clj/editor/editor_extensions/ui_docs.clj

```diff
@@ -0,0 +1,718 @@
+        input-with-variant-props))
+
+;; endregion
+
+;; region component definitions
+
+(defn- props-doc-html [props]
+  (lua-completion/args-doc-html
+    (map #(update % :name name) props)))
+
+(s/def :editor.editor-extensions.components.component/name string?)
+(s/def ::props (s/coll-of ::prop))
+(s/def ::description string?)
+(s/def ::component
+  (s/keys :req-un [:editor.editor-extensions.components.component/name ::props ::description]))
+
+(defn ^{:arglists '([name & {:keys [props description]}])} component
+  "Construct a component definition map
+
+  Args:
+    name    string name of the component, should be snake_cased
+
+  Kv-args:
+    :props          required, a list of props of the component
+    :description    required, markdown documentation string
+
+  Returns a component definition map with :name, :props and :description keys"
+  [name & {:as m}]
+  {:post [(s/assert ::component %)]}
+  (assoc m :name name))
+
+(def horizontal-component
+  (component
+    "horizontal"
+    :description "Layout container that places its children in a horizontal row one after another"
+    :props list-props))
+
+(def vertical-component
+  (component
+    "vertical"
+    :description "Layout container that places its children in a vertical column one after another"
+    :props list-props))
+
+(def grid-component
+  (component
+    "grid"
+    :description "Layout container that places its children in a 2D grid"
+    :props grid-props))
+
+(def separator-component
+  (component
+    "separator"
+    :description "Thin line for visual content separation, by default horizontal and aligned to center"
+    :props separator-props))
+
+(def scroll-component
+  (component
+    "scroll"
+    :description "Layout container that optionally shows scroll bars if child contents overflow the assigned bounds"
+    :props scroll-props))
+
+(def label-component
+  (component
+    "label"
+    :description "Non-resizeable text label"
+    :props label-props))
+
+(def paragraph-component
+  (component
+    "paragraph"
+    :description "Typography component representing a paragraph of text"
+    :props paragraph-props))
+
+(def heading-component
+  (component
+    "heading"
+    :description "Typography component representing a heading"
+    :props heading-props))
+
+(def icon-component
+  (component
+    "icon"
+    :description "An icon of a fixed size"
+    :props icon-props))
+
+(def button-component
+  (component
+    "button"
+    :props button-props
+    :description "Button with a label and/or an icon"))
+
+(def check-box-component
+  (component
+    "check_box"
+    :props check-box-props
+    :description "Check box with a label"))
+
+(def select-box-component
+  (component
+    "select_box"
+    :props select-box-props
+    :description "Dropdown select box with an array of options"))
+
+(def text-field-component
+  (component
+    "text_field"
+    :props text-field-props
+    :description "Single-line text field, reports changes on typing"))
```

Comment:
I thought we agreed to not do it 😅 
As we discussed, the components have different behavior and are intended for different purposes. You can't reproduce a `text_field` by adding an `on_text_changed` callback to the `string_field`, because it will have behaviors that you don't want. Specifically, you don't want to have "commit on Enter" and "rollback to previous value on Escape" behaviors in a `text_field`. You want these different components in different scenarios: 
- `text_field` for immediate feedback on any typing, e.g. for immediate validation of new file names.
- `value_field` and friends for feedback on explicit commit action, e.g. for number form inputs where intermediate text is often invalid (like `-`, `1.`), but immediate feedback is only distracting.